mne/beamformer/tests/test_dics.py

```diff
@@ -26,34 +36,400 @@
+
+    # Test if wrong channel selection is detected in application of filter
+    evoked_ch = cp.deepcopy(evoked)
+    evoked_ch.pick_channels(evoked_ch.ch_names[:-1])
+    raises(ValueError, apply_dics, evoked_ch, filters)
+
+    # Try different types of forward models
+    for fwd in [fwd_free, fwd_surf, fwd_fixed]:
+        filters = make_dics(evoked.info, fwd, csd20, label=label, reg=reg)
+
+        # Apply filters to evoked
+        stc = apply_dics(evoked, filters)
+        stc = (stc ** 2).mean()
+        assert np.argmax(stc.data) == source_ind
+
+        # Apply filters to epochs
+        stcs = apply_dics_epochs(epochs, filters)
+        stc = (stcs[0] ** 2).mean()
+        assert np.argmax(stc.data) == source_ind
+
+    # Try picking different orientations
+    for pick_ori in [None, 'normal', 'max-power']:
+        filters = make_dics(evoked.info, fwd_surf, csd20, label=label, reg=reg,
+                            pick_ori=pick_ori)
+
+        # Apply to evoked
+        stc = apply_dics(evoked, filters)
+        stc = (stc ** 2).mean()
+        assert np.argmax(stc.data) == source_ind
+
+        # Apply to epochs
+        stcs = apply_dics_epochs(epochs, filters)
+        stc = (stcs[0] ** 2).mean()
+        assert np.argmax(stc.data) == source_ind
+
+    # Test using a real-valued filter
+    filters_real = make_dics(evoked.info, fwd_surf, csd20, label=label,
+                             reg=reg, real_filter=True)
+    assert not np.iscomplexobj(filters_real['weights'])
+    stc = apply_dics(evoked, filters_real)
+    stc = (stc ** 2).mean()
+    assert np.argmax(stc.data) == source_ind
+
+    # Test whether projections are applied, by adding a custom projection
+    filters_noproj = make_dics(evoked.info, fwd_surf, csd20, label=label)
+    stc_noproj = apply_dics(evoked, filters_noproj)
+    evoked_proj = evoked.copy()
+    p = compute_proj_evoked(evoked_proj, n_grad=1, n_mag=0, n_eeg=0)
+    proj_matrix = make_projector(p, evoked_proj.ch_names)[0]
+    evoked_proj.info['projs'] += p
+    filters_proj = make_dics(evoked_proj.info, fwd_surf, csd20, label=label)
+    assert_array_equal(filters_proj['proj'], proj_matrix)
+    stc_proj = apply_dics(evoked_proj, filters_proj)
+    assert np.any(np.not_equal(stc_noproj.data, stc_proj.data))
+
+    # Test detecting incompatible projections
+    filters_proj['proj'] = filters_proj['proj'][:-1, :-1]
+    raises(ValueError, apply_dics, evoked_proj, filters_proj)
+
+    # Test returning a generator
+    stcs = apply_dics_epochs(epochs, filters, return_generator=False)
+    stcs_gen = apply_dics_epochs(epochs, filters, return_generator=True)
+    assert_array_equal(stcs[0].data, advance_iterator(stcs_gen).data)
+
+    # Test computing timecourses on a volume source space
+    filters_vol = make_dics(evoked.info, fwd_vol, csd20, reg=reg)
+    stc = apply_dics(evoked, filters_vol)
+    stc = (stc ** 2).mean()
+    assert np.argmax(stc.data) == 3851  # TODO: don't make this hard coded
```

Comment:
Are any of these time consuming? If so the `pytest.mark` as slow would be useful

Reply:
Each of the steps are pretty fast, but the test in total is starting to take time. So many different switches to test. I'm guessing the LCMV tests have a similar problem.